marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java

```diff
@@ -45,24 +48,61 @@
     /**
      *
      * @param hubClient
-     * @param taskId
      * @param schema
      * @param params contains all the params provided by Spark, which will include all connector-specific properties
      */
-    public HubDataWriter(HubClient hubClient, long taskId, StructType schema, Map<String, String> params) {
+    public HubDataWriter(HubClient hubClient, StructType schema, Map<String, String> params) {
         this.records = new ArrayList<>();
         this.schema = schema;
         this.batchSize = params.containsKey("batchsize") ? Integer.parseInt(params.get("batchsize")) : 100;
 
-        final String apiModulePath = params.containsKey("apipath") ? params.get("apipath") : "/data-hub/5/data-services/ingestion/bulkIngester.api";
-        logger.info("Will write to endpoint defined by: " + apiModulePath);
-        this.loader = InputEndpoint.on(
-            hubClient.getStagingClient(),
-            hubClient.getModulesClient().newTextDocumentManager().read(apiModulePath, new StringHandle())
-        ).bulkCaller();
+        String ingestEndpointParams = params.get("ingestendpointparams");
+        try {
+            JSONObject jsonObject = new JSONObject(ingestEndpointParams);
 
-        loader.setWorkUnit(new ByteArrayInputStream(("{\"taskId\":" + taskId + "}").getBytes()));
-        loader.setEndpointState(new ByteArrayInputStream(("{\"next\":" + 0 + ", \"uriprefix\":\"" + params.get("uriprefix") + "\"}").getBytes()));
+            if(jsonObject.get("apiPath") == null || jsonObject.get("apiPath").toString().length()==0) {
+                if((jsonObject.get("endpointState") != null && jsonObject.get("endpointState").toString().length()>0) ||
+                    (jsonObject.get("workUnit")!=null && jsonObject.get("workUnit").toString().length()>0))
+                {
+                    throw new RuntimeException("Cannot override default endpointState and/or workUnit. Please provide custom endpoint.");
+                }
+            }
+            final String apiModulePath = (jsonObject.get("apiPath")!=null && jsonObject.get("apiPath").toString().length()>0) ?
```

Comment:
There's a lot going on here, and generally, I like to use private/protected methods to break up long methods so that they're easier to read and maintain. I prefer private, though I'll use protected when I want to write unit tests against the logic.

I see the following steps in this code, where each can become a new private method:

1. Use the user params to determine the apiPath, the workUnit, and endpointState (and throw a validation error as needed)
2. Use the output of #1 to create a BulkInputCaller

I think the signature of the first private method would be "private JsonNode determineIngestionEndpointParams(Map)". If the user has defined "ingestendpointparams", then use ObjectMapper to parse that into a JsonNode, and then validate it. Else, use objectMapper.createObjectNode() and populate apiPath based on the default API path, and then populate workUnit based on uriPrefix (and then @SameeraPriyathamTadikonda  would add more logic to this for collections/permissions/etc). 

That first method would likely be protected so that you could easily write unit tests against it without having to involve everything else in this class. 

The second private method would have a signature of "private BulkInputCaller buildBulkInputCaller(JsonNode ingestionParams)". 

Each private method can then have a try/catch as needed to ensure that if an error is thrown, the appropriate context is provided in the error message.